ast.go

```diff
@@ -67,10 +69,30 @@ func (*Update) iStatement()         {}
 func (*Grant) iStatement()          {}
 func (*Revoke) iStatement()         {}
 
+// ReadStatementResolver resolves Tableland Custom Functions for a read statement.
+type ReadStatementResolver interface {
+	// GetBlockNumber returns the last known block number for the provided chainID. If the chainID isn't known,
+	// it returns (0, false).
+	GetBlockNumber(chainID int64) (int64, bool)
+}
+
+// WriteStatementResolver resolves Tableland Custom Functions for a write statement.
+type WriteStatementResolver interface {
+	// GetTxnHash returns the transaction hash of the transaction containing the query being processed.
+	GetTxnHash() string
+
+	// GetBlockNumber returns the block number of the block containing query being processed.
+	GetBlockNumber() int64
+}
+
 // ReadStatement is any SELECT statement or UNION statement.
 type ReadStatement interface {
 	iReadStatement()
 	iStatement()
+
+	// Resolve returns a string representation with custom function nodes resolved to the values
+	// passed by resolver.
+	Resolve(ReadStatementResolver) (string, error)
```

Comment:
All `ReadStatement`s must implement this method. Those statements are `Select` and `CompoundSelect`